File: artifacts/claimflow-showcase/render_video.m
#import <Foundation/Foundation.h>
#import <AVFoundation/AVFoundation.h>
#import <CoreGraphics/CoreGraphics.h>
#import <CoreVideo/CoreVideo.h>
#import <ImageIO/ImageIO.h>

static CGImageRef LoadImage(NSURL *url) {
    CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
    if (!source) return NULL;
    CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
    CFRelease(source);
    return image;
}

static void DrawImage(CGContextRef context, CGImageRef image, int width, int height, CGFloat scale, CGFloat alpha) {
    CGFloat drawWidth = width * scale;
    CGFloat drawHeight = height * scale;
    CGRect rect = CGRectMake((width - drawWidth) / 2.0, (height - drawHeight) / 2.0, drawWidth, drawHeight);
    CGContextSaveGState(context);
    CGContextSetAlpha(context, alpha);
    CGContextDrawImage(context, rect, image);
    CGContextRestoreGState(context);
}

static CVPixelBufferRef MakeFrame(CVPixelBufferPoolRef pool, CGImageRef current, CGImageRef next, double progress, double transition, int width, int height) {
    CVPixelBufferRef buffer = NULL;
    if (CVPixelBufferPoolCreatePixelBuffer(NULL, pool, &buffer) != kCVReturnSuccess) return NULL;
    CVPixelBufferLockBaseAddress(buffer, 0);
    void *base = CVPixelBufferGetBaseAddress(buffer);
    size_t bytesPerRow = CVPixelBufferGetBytesPerRow(buffer);
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = CGBitmapContextCreate(base, width, height, 8, bytesPerRow, colorSpace, kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Little);
    CGColorSpaceRelease(colorSpace);
    if (!context) {
        CVPixelBufferUnlockBaseAddress(buffer, 0);
        CVPixelBufferRelease(buffer);
        return NULL;
    }
    CGContextSetRGBFillColor(context, 0.984, 0.98, 0.965, 1.0);
    CGContextFillRect(context, CGRectMake(0, 0, width, height));
    CGContextTranslateCTM(context, 0, height);
    CGContextScaleCTM(context, 1, -1);
    CGFloat currentScale = 1.0 + progress * 0.018;
    if (transition > 0 && next) {
        DrawImage(context, current, width, height, currentScale, 1.0 - transition);
        DrawImage(context, next, width, height, 1.0 + transition * 0.004, transition);
    } else {
        DrawImage(context, current, width, height, currentScale, 1.0);
    }
    CGContextRelease(context);
    CVPixelBufferUnlockBaseAddress(buffer, 0);
    return buffer;
}

static BOOL RenderSilent(NSURL *root, NSArray<NSDictionary *> *scenes, NSURL *output, NSError **error) {
    [[NSFileManager defaultManager] removeItemAtURL:output error:nil];
    const int width = 1920, height = 1080, fps = 5;
    AVAssetWriter *writer = [[AVAssetWriter alloc] initWithURL:output fileType:AVFileTypeQuickTimeMovie error:error];
    if (!writer) return NO;
    NSDictionary *settings = @{AVVideoCodecKey:AVVideoCodecTypeAppleProRes422LT, AVVideoWidthKey:@(width), AVVideoHeightKey:@(height)};
    AVAssetWriterInput *input = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:settings];
    input.expectsMediaDataInRealTime = NO;
    NSDictionary *attrs = @{(NSString *)kCVPixelBufferPixelFormatTypeKey:@(kCVPixelFormatType_32BGRA), (NSString *)kCVPixelBufferWidthKey:@(width), (NSString *)kCVPixelBufferHeightKey:@(height)};
    AVAssetWriterInputPixelBufferAdaptor *adaptor = [AVAssetWriterInputPixelBufferAdaptor assetWriterInputPixelBufferAdaptorWithAssetWriterInput:input sourcePixelBufferAttributes:attrs];
    if (![writer canAddInput:input]) return NO;
    [writer addInput:input];
    if (![writer startWriting]) { if (error) *error = writer.error; return NO; }
    [writer startSessionAtSourceTime:kCMTimeZero];
    NSMutableArray *images = [NSMutableArray array];
    for (NSInteger i = 0; i < scenes.count; i++) {
        NSString *name = [NSString stringWithFormat:@"frames/scene-%02ld.png", (long)i + 1];
        CGImageRef image = LoadImage([root URLByAppendingPathComponent:name]);
        if (!image) return NO;
        [images addObject:(__bridge id)image];
        CGImageRelease(image);
    }
    int64_t absoluteFrame = 0;
    int64_t transitionFrames = MAX((int64_t)2, (int64_t)llround(0.8 * fps));
    for (NSInteger sceneIndex = 0; sceneIndex < scenes.count; sceneIndex++) {
        NSDictionary *scene = scenes[sceneIndex];
        int64_t sceneFrames = (int64_t)llround([scene[@"duration"] doubleValue] * fps);
        CGImageRef current = (__bridge CGImageRef)images[sceneIndex];
        CGImageRef next = sceneIndex + 1 < images.count ? (__bridge CGImageRef)images[sceneIndex + 1] : NULL;
        for (int64_t localFrame = 0; localFrame < sceneFrames; localFrame++) {
            while (!input.readyForMoreMediaData) [NSThread sleepForTimeInterval:0.002];
            double progress = (double)localFrame / MAX((double)sceneFrames - 1, 1);
            double transition = 0;
            if (next && localFrame >= sceneFrames - transitionFrames) {
                transition = (double)(localFrame - (sceneFrames - transitionFrames)) / MAX((double)transitionFrames - 1, 1);
            }
            CVPixelBufferRef buffer = MakeFrame(adaptor.pixelBufferPool, current, next, progress, transition, width, height);
            if (!buffer) return NO;
            BOOL ok = [adaptor appendPixelBuffer:buffer withPresentationTime:CMTimeMake(absoluteFrame, fps)];
            CVPixelBufferRelease(buffer);
            if (!ok) { if (error) *error = writer.error; return NO; }
            absoluteFrame++;
        }
    }
    [input markAsFinished];
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
    [writer finishWritingWithCompletionHandler:^{ dispatch_semaphore_signal(sem); }];
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
    if (writer.status != AVAssetWriterStatusCompleted) { if (error) *error = writer.error; return NO; }
    return YES;
}

static BOOL MuxAudio(NSURL *videoURL, NSURL *audioURL, NSURL *outputURL, NSError **error) {
    [[NSFileManager defaultManager] removeItemAtURL:outputURL error:nil];
    AVURLAsset *video = [AVURLAsset URLAssetWithURL:videoURL options:nil];
    AVURLAsset *audio = [AVURLAsset URLAssetWithURL:audioURL options:nil];
    AVMutableComposition *composition = [AVMutableComposition composition];
    AVAssetTrack *sourceVideo = [[video tracksWithMediaType:AVMediaTypeVideo] firstObject];
    AVMutableCompositionTrack *videoTrack = [composition addMutableTrackWithMediaType:AVMediaTypeVideo preferredTrackID:kCMPersistentTrackID_Invalid];
    if (!sourceVideo || !videoTrack) return NO;
    if (![videoTrack insertTimeRange:CMTimeRangeMake(kCMTimeZero, video.duration) ofTrack:sourceVideo atTime:kCMTimeZero error:error]) return NO;
    videoTrack.preferredTransform = sourceVideo.preferredTransform;
    AVAssetTrack *sourceAudio = [[audio tracksWithMediaType:AVMediaTypeAudio] firstObject];
    AVMutableCompositionTrack *audioTrack = [composition addMutableTrackWithMediaType:AVMediaTypeAudio preferredTrackID:kCMPersistentTrackID_Invalid];
    if (sourceAudio && audioTrack) {
        CMTime audioDuration = CMTimeCompare(audio.duration, video.duration) < 0 ? audio.duration : video.duration;
        if (![audioTrack insertTimeRange:CMTimeRangeMake(kCMTimeZero, audioDuration) ofTrack:sourceAudio atTime:kCMTimeZero error:error]) return NO;
    }
    AVAssetExportSession *exporter = [[AVAssetExportSession alloc] initWithAsset:composition presetName:AVAssetExportPresetPassthrough];
    exporter.outputURL = outputURL;
    exporter.outputFileType = AVFileTypeQuickTimeMovie;
    exporter.shouldOptimizeForNetworkUse = YES;
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
    [exporter exportAsynchronouslyWithCompletionHandler:^{ dispatch_semaphore_signal(sem); }];
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
    if (exporter.status != AVAssetExportSessionStatusCompleted) { if (error) *error = exporter.error; return NO; }
    return YES;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *rootPath = argc > 1 ? [NSString stringWithUTF8String:argv[1]] : [[NSFileManager defaultManager] currentDirectoryPath];
        NSURL *root = [NSURL fileURLWithPath:rootPath isDirectory:YES];
        NSData *data = [NSData dataWithContentsOfURL:[root URLByAppendingPathComponent:@"storyboard.json"]];
        NSError *error = nil;
        NSArray<NSDictionary *> *scenes = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
        if (!scenes || error) { fprintf(stderr, "Could not read storyboard: %s\n", error.localizedDescription.UTF8String); return 1; }
        double total = 0; for (NSDictionary *scene in scenes) total += [scene[@"duration"] doubleValue];
        if (fabs(total - 90.0) > 0.001) { fprintf(stderr, "Storyboard is %.2f seconds, expected 90\n", total); return 1; }
        NSURL *silent = [root URLByAppendingPathComponent:@"claimflow-showcase-silent.mov"];
        NSURL *audio = [root URLByAppendingPathComponent:@"narration.aiff"];
        NSURL *output = [root URLByAppendingPathComponent:@"claimflow-ai-product-story-90s.mov"];
        printf("Rendering %ld scenes across 90 seconds…\n", (long)scenes.count);
        if (!RenderSilent(root, scenes, silent, &error)) { fprintf(stderr, "Silent render failed: %s | %s\n", error.localizedDescription.UTF8String, error.userInfo.description.UTF8String); return 1; }
        if (!MuxAudio(silent, audio, output, &error)) { fprintf(stderr, "Audio export failed: %s | %s\n", error.localizedDescription.UTF8String, error.userInfo.description.UTF8String); return 1; }
        printf("Created %s\n", output.path.UTF8String);
    }
    return 0;
}

File: artifacts/claimflow-showcase/render_video.swift
import Foundation
import AVFoundation
import CoreGraphics
import CoreVideo
import ImageIO

struct Scene: Codable {
    let image: String
    let title: String
    let subtitle: String
    let duration: Double
}

enum RenderError: Error, CustomStringConvertible {
    case message(String)
    var description: String {
        switch self { case .message(let text): return text }
    }
}

func loadImage(_ url: URL) throws -> CGImage {
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
          let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
        throw RenderError.message("Could not load image at \(url.path)")
    }
    return image
}

func drawImage(_ image: CGImage, in context: CGContext, width: Int, height: Int, scale: CGFloat, alpha: CGFloat) {
    let drawWidth = CGFloat(width) * scale
    let drawHeight = CGFloat(height) * scale
    let x = (CGFloat(width) - drawWidth) / 2
    let y = (CGFloat(height) - drawHeight) / 2
    context.saveGState()
    context.setAlpha(alpha)
    context.draw(image, in: CGRect(x: x, y: y, width: drawWidth, height: drawHeight))
    context.restoreGState()
}

func makePixelBuffer(pool: CVPixelBufferPool, current: CGImage, next: CGImage?, progress: Double, transition: Double, width: Int, height: Int) throws -> CVPixelBuffer {
    var maybeBuffer: CVPixelBuffer?
    let status = CVPixelBufferPoolCreatePixelBuffer(nil, pool, &maybeBuffer)
    guard status == kCVReturnSuccess, let buffer = maybeBuffer else {
        throw RenderError.message("Could not allocate video frame")
    }
    CVPixelBufferLockBaseAddress(buffer, [])
    defer { CVPixelBufferUnlockBaseAddress(buffer, []) }
    guard let base = CVPixelBufferGetBaseAddress(buffer) else {
        throw RenderError.message("Missing pixel buffer base address")
    }
    let bytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
    guard let context = CGContext(
        data: base,
        width: width,
        height: height,
        bitsPerComponent: 8,
        bytesPerRow: bytesPerRow,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
    ) else {
        throw RenderError.message("Could not create frame drawing context")
    }
    context.setFillColor(CGColor(red: 0.984, green: 0.98, blue: 0.965, alpha: 1))
    context.fill(CGRect(x: 0, y: 0, width: width, height: height))
    context.translateBy(x: 0, y: CGFloat(height))
    context.scaleBy(x: 1, y: -1)
    let currentScale = 1.0 + CGFloat(progress) * 0.018
    if transition > 0, let nextImage = next {
        drawImage(current, in: context, width: width, height: height, scale: currentScale, alpha: CGFloat(1.0 - transition))
        drawImage(nextImage, in: context, width: width, height: height, scale: 1.0 + CGFloat(transition) * 0.004, alpha: CGFloat(transition))
    } else {
        drawImage(current, in: context, width: width, height: height, scale: currentScale, alpha: 1)
    }
    return buffer
}

func renderSilentVideo(root: URL, scenes: [Scene], output: URL) throws {
    try? FileManager.default.removeItem(at: output)
    let width = 1920
    let height = 1080
    let fps: Int32 = 30
    let writer = try AVAssetWriter(outputURL: output, fileType: .mp4)
    let settings: [String: Any] = [
        AVVideoCodecKey: AVVideoCodecType.h264,
        AVVideoWidthKey: width,
        AVVideoHeightKey: height,
        AVVideoCompressionPropertiesKey: [
            AVVideoAverageBitRateKey: 8_000_000,
            AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel
        ]
    ]
    let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
    input.expectsMediaDataInRealTime = false
    let attributes: [String: Any] = [
        kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
        kCVPixelBufferWidthKey as String: width,
        kCVPixelBufferHeightKey as String: height
    ]
    let adaptor = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: input, sourcePixelBufferAttributes: attributes)
    guard writer.canAdd(input) else { throw RenderError.message("Video input is not supported") }
    writer.add(input)
    guard writer.startWriting() else { throw writer.error ?? RenderError.message("Video writer failed to start") }
    writer.startSession(atSourceTime: .zero)
    guard let pool = adaptor.pixelBufferPool else { throw RenderError.message("Pixel buffer pool is unavailable") }
    let frames = try scenes.enumerated().map { index, scene in
        try loadImage(root.appendingPathComponent("frames/scene-\(String(format: "%02d", index + 1)).png"))
    }
    var absoluteFrame: Int64 = 0
    let transitionSeconds = 0.65
    for (sceneIndex, scene) in scenes.enumerated() {
        let sceneFrameCount = Int64((scene.duration * Double(fps)).rounded())
        let transitionFrames = Int64(transitionSeconds * Double(fps))
        for localFrame in 0..<sceneFrameCount {
            while !input.isReadyForMoreMediaData { Thread.sleep(forTimeInterval: 0.002) }
            let progress = Double(localFrame) / Double(max(sceneFrameCount - 1, 1))
            let transition: Double
            if sceneIndex < scenes.count - 1 && localFrame >= sceneFrameCount - transitionFrames {
                transition = Double(localFrame - (sceneFrameCount - transitionFrames)) / Double(max(transitionFrames - 1, 1))
            } else {
                transition = 0
            }
            let buffer = try makePixelBuffer(
                pool: pool,
                current: frames[sceneIndex],
                next: sceneIndex < frames.count - 1 ? frames[sceneIndex + 1] : nil,
                progress: progress,
                transition: transition,
                width: width,
                height: height
            )
            let time = CMTime(value: absoluteFrame, timescale: fps)
            if !adaptor.append(buffer, withPresentationTime: time) {
                throw writer.error ?? RenderError.message("Could not append video frame")
            }
            absoluteFrame += 1
        }
    }
    input.markAsFinished()
    let semaphore = DispatchSemaphore(value: 0)
    writer.finishWriting { semaphore.signal() }
    semaphore.wait()
    guard writer.status == .completed else {
        throw writer.error ?? RenderError.message("Video writer did not complete")
    }
}

func muxAudio(videoURL: URL, audioURL: URL, outputURL: URL) throws {
    try? FileManager.default.removeItem(at: outputURL)
    let videoAsset = AVURLAsset(url: videoURL)
    let audioAsset = AVURLAsset(url: audioURL)
    let composition = AVMutableComposition()
    guard let sourceVideo = videoAsset.tracks(withMediaType: .video).first,
          let compositionVideo = composition.addMutableTrack(withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid) else {
        throw RenderError.message("Could not load the silent video track")
    }
    try compositionVideo.insertTimeRange(CMTimeRange(start: .zero, duration: videoAsset.duration), of: sourceVideo, at: .zero)
    compositionVideo.preferredTransform = sourceVideo.preferredTransform
    if let sourceAudio = audioAsset.tracks(withMediaType: .audio).first,
       let compositionAudio = composition.addMutableTrack(withMediaType: .audio, preferredTrackID: kCMPersistentTrackID_Invalid) {
        let audioDuration = CMTimeMinimum(audioAsset.duration, videoAsset.duration)
        try compositionAudio.insertTimeRange(CMTimeRange(start: .zero, duration: audioDuration), of: sourceAudio, at: .zero)
    }
    guard let exporter = AVAssetExportSession(asset: composition, presetName: AVAssetExportPresetHighestQuality) else {
        throw RenderError.message("Could not create the final video exporter")
    }
    exporter.outputURL = outputURL
    exporter.outputFileType = .mp4
    exporter.shouldOptimizeForNetworkUse = true
    let semaphore = DispatchSemaphore(value: 0)
    exporter.exportAsynchronously { semaphore.signal() }
    semaphore.wait()
    guard exporter.status == .completed else {
        throw exporter.error ?? RenderError.message("Final video export did not complete")
    }
}

let root = URL(fileURLWithPath: CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : FileManager.default.currentDirectoryPath)
let storyboardURL = root.appendingPathComponent("storyboard.json")
let silentURL = root.appendingPathComponent("claimflow-showcase-silent.mp4")
let narrationURL = root.appendingPathComponent("narration.aiff")
let outputURL = root.appendingPathComponent("claimflow-ai-product-story-90s.mp4")

do {
    let scenes = try JSONDecoder().decode([Scene].self, from: Data(contentsOf: storyboardURL))
    let total = scenes.reduce(0) { $0 + $1.duration }
    guard abs(total - 90.0) < 0.001 else { throw RenderError.message("Storyboard must total exactly 90 seconds; found \(total)") }
    print("Rendering \(scenes.count) scenes across \(Int(total)) seconds…")
    try renderSilentVideo(root: root, scenes: scenes, output: silentURL)
    try muxAudio(videoURL: silentURL, audioURL: narrationURL, outputURL: outputURL)
    print("Created \(outputURL.path)")
} catch {
    fputs("Render failed: \(error)\n", stderr)
    exit(1)
}
